### PR TITLE
WI #1864 Add peformance test with CFG/DFA

### PR DESCRIPTION
**Pull request by mayanje**
**What have been done:**
Add two classes in TypeCobol.Analysis.Test

- PerformanceWithCfg
- PerformanceWithCfgDfa

Both inherit from class TypeCobol.Test.Performance

### CODE REVIEW

File: TypeCobol.Analysis.Test/PerformanceWithCfg.cs [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+            document.Init(fullPath, options, format, copiesFolder, analyzerProvider);
+            document.Parse(fullPath);
+            return document;
+        }
+
+        [TestMethod]
+        [TestCategory("Performance")]
+        [TestProperty("Time", "fast")]
+        //[Ignore]
+        public new void AntlrPerformanceProfiler()
```
Comment (smedilol) [inline TypeCobol.Analysis.Test/PerformanceWithCfg.cs:45]: You can remove AntlrPerformanceProfiler test, because CodeElement/Antlr parsing is not affected by CFG.

File: TypeCobol.Analysis.Test/PerformanceWithCfgDfa.cs [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+            document.Init(fullPath, options, format, copiesFolder, analyzerProvider);
+            document.Parse(fullPath);
+            return document;
+        }
+
+        [TestMethod]
+        [TestCategory("Performance")]
+        [TestProperty("Time", "fast")]
+        //[Ignore]
+        public new void AntlrPerformanceProfiler()
```
Comment (smedilol) [inline TypeCobol.Analysis.Test/PerformanceWithCfgDfa.cs:45]: Same than CFG, you can remove AntlrPerformanceProfiler test.

**Review by smedilol — CHANGES_REQUESTED** [review]

File: TypeCobol.Analysis.Test/PerformanceWithCfgDfa.cs (hunk 2 above)
Comment (mayanje) [reply TypeCobol.Analysis.Test/PerformanceWithCfgDfa.cs]: done in af9e8cb

File: TypeCobol.Analysis.Test/PerformanceWithCfg.cs (hunk 1 above)
Comment (mayanje) [reply TypeCobol.Analysis.Test/PerformanceWithCfg.cs]: done in af9e8cb

File: TypeCobol.Test/Parser/Performance/Performance.cs [hunk 3]
```diff
@@ -368,8 +368,21 @@ public void Part3_FullParsing_TC_DeepTypes()
...
-
+        /// <summary>
+        /// Method for parsinga document.
+        /// </summary>
+        /// <param name="fullPath"></param>
+        /// <param name="options"></param>
+        /// <param name="format"></param>
+        /// <param name="copiesFolder"></param>
+        /// <returns></returns>
+        protected virtual TypeCobol.Parser parseDocument(string fullPath, TypeCobolOptions options, TypeCobol.Compiler.DocumentFormat format, string[] copiesFolder)
```
Comment (fm-117) [inline TypeCobol.Test/Parser/Performance/Performance.cs:379]: `ParseDocument` to match PascalCase convention.

About the pattern used, you can improve it and avoid these methods marked as `new` and calling base code. You need 4 classes :
- Performance : abstract base class, contains all the tests and an abstract ParseDocument method to initialize the parser
- 3 concrete classes, 1 per mode
  - NoCfg
  - Cfg
  - Dfa

Each concrete class would implement the ParseDocument method to match its own needs and you can override the Antlr test in Cfg and Dfa to make it empty.

**Review by fm-117 — CHANGES_REQUESTED** [review]

File: TypeCobol.Test/Parser/Performance/Performance.cs (hunk 3 above)
Comment (fm-117) [reply TypeCobol.Test/Parser/Performance/Performance.cs:379]: Capture
As you can see, `TestMethod1`  is defined once but translates into multiple tests 👍

Comment (mayanje) [reply TypeCobol.Test/Parser/Performance/Performance.cs:379]: Sorry for PascalCase convention it's the inverse in Java, I will do the change. Also I'll try your pattern because it's true if it work like that it's fine.

Comment (mayanje) [reply TypeCobol.Test/Parser/Performance/Performance.cs:379]: At first time I did what to suggest, the Performance base class was not abstract, it didn't work, so maybe with the abstract keyword it will work

Comment (mayanje) [reply TypeCobol.Test/Parser/Performance/Performance.cs]: This method seems not working if tests classes are in different assemblies.

**Review by fm-117 — DISMISSED** [review]
When we'll upgrade the test framework, we'll be able to remove the test duplication, but as for now we have to keep the duplicated `new` methods.

File: TypeCobol.Test/Parser/Performance/Performance.cs [hunk 4]
```diff
@@ -368,8 +368,21 @@ public void Part3_FullParsing_TC_DeepTypes()
...
-
+        /// <summary>
+        /// Method for parsinga document.
+        /// </summary>
+        /// <param name="fullPath"></param>
+        /// <param name="options"></param>
+        /// <param name="format"></param>
+        /// <param name="copiesFolder"></param>
+        /// <returns></returns>
+        protected virtual TypeCobol.Parser ParseDocument(string fullPath, TypeCobolOptions options, TypeCobol.Compiler.DocumentFormat format, string[] copiesFolder)
```
Comment (smedilol) [inline TypeCobol.Test/Parser/Performance/Performance.cs:379]: This method seems only for FullParsing. Does it also include Incremental ?

**Review by rooksdo — DISMISSED** [review]

File: TypeCobol.Test/Parser/Performance/Performance.cs (hunk 4 above)
Comment (mayanje) [reply TypeCobol.Test/Parser/Performance/Performance.cs:395]: You are right, done in 3b1d62e

**Review by smedilol — APPROVED** [review]

File: TypeCobol.Analysis.Test/PerformanceWithCfg.cs [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+            var document = new TypeCobol.Parser();
+            document.Init(fullPath, options, format, copiesFolder, MyAnalyzerProvider());
+            document.Parse(fullPath);
+            return document;
+        }
+
+        protected override CompilationProject CreateCompilationProject(string projectName, string rootDirectory, string[] fileExtensions, DocumentFormat documentFormat, TypeCobolOptions compilationOptions, IAnalyzerProvider analyzerProvider)
+        {
+            return base.CreateCompilationProject(projectName, rootDirectory, fileExtensions, documentFormat, compilationOptions, MyAnalyzerProvider());
+        }
```
Comment (fm-117) [inline TypeCobol.Analysis.Test/PerformanceWithCfg.cs:52]: Ok so it looks like `ParseDocument` and `CreateCompilationProject` are respectively used for FullParsing and Incremental. You can improve factorization by removing these 2 methods (as they only differ at the type of analyzer used in derived classes) and replace them with something like :
```csharp
protected abstract void AddAnalyzers(CompositeAnalyzerProvider analyzerProvider);
```
in the base class `Performance`. No need to change anything from original code except to create a `CompositeAnalyzerProvider`, call `AddAnalyzer` and pass the instance either on the `Parser.Init` or into the `CompilationProject`.

**Review by fm-117 — CHANGES_REQUESTED** [review]

**Review by rooksdo — APPROVED** [review]

**Review by fm-117 — APPROVED** [review]

File: TypeCobol.Analysis.Test/PerformanceWithCfg.cs (hunk 5 above)
Comment (mayanje) [reply TypeCobol.Analysis.Test/PerformanceWithCfg.cs]: Done in 09d4586

**Review by smedilol — APPROVED** [review]

**Review by rooksdo — APPROVED** [review]
